gradle-consistent-versions-idea-plugin/src/main/java/com/palantir/gradle/versions/intellij/VersionCompletionContributor.java

```diff
@@ -104,52 +112,82 @@ private void addLoadingElement(CompletionResultSet sortedResultSet) {
+
+        List<PackageInRepo> allPackages = RepositoryLoader.loadRepositories(project).stream()
+                .map(url -> new PackageInRepo(dependencyInfo.group(), dependencyInfo.dependencyName(), url))
+                .collect(Collectors.toList());
+        addToResults(sortedResultSet, allPackages, dependencyInfo);
     }
 
-    private List<PackageInRepo> collectGroupAndDeps(
-            Project project, DependencyGroup group, DependencyName dependencyName) {
+    private List<PackageInRepo> collectPackageInRepo(Project project, DependencyInfo dependencyInfo) {
 
-        String dependencyNamePrefix = dependencyName.name().replace("*", "");
+        String dependencyNamePrefix = dependencyInfo.dependencyName().name().replace("*", "");
         return StreamEx.of(RepositoryLoader.loadRepositories(project))
-                .flatMap(url -> StreamEx.of(
-                                groupPartOrPackageNameExplorer.getCancelableGroupPartOrPackageName(group, url))
+                .flatMap(url -> StreamEx.of(groupPartOrPackageNameExplorer.getCancelableGroupPartOrPackageName(
+                                dependencyInfo.group(), url))
                         .filter(pkgName -> pkgName.name().startsWith(dependencyNamePrefix))
                         .map(pkgName -> new SimpleEntry<>(url, pkgName)))
                 .map(entry -> {
                     RepositoryUrl url = entry.getKey();
                     GroupPartOrPackageName pkgName = entry.getValue();
                     DependencyName depName = DependencyName.of(pkgName.name());
-                    return new PackageInRepo(group, depName, url);
+                    return new PackageInRepo(dependencyInfo.group(), depName, url);
                 })
                 .toList();
     }
 
-    private void addToResults(CompletionResultSet resultSet, List<PackageInRepo> groupAndDeps) {
-        Map<DependencyVersion, AtomicInteger> versionCounts = StreamEx.of(groupAndDeps)
-                .flatMap(groupAndDep ->
-                        versionExplorer
-                                .getVersions(groupAndDep, CompletionRefreshUtil.refreshOnceSupplier()::get)
-                                .stream())
-                .collect(Collectors.toConcurrentMap(
-                        Function.identity(), v -> new AtomicInteger(1), (existingCount, newCount) -> {
-                            existingCount.addAndGet(newCount.get());
-                            return existingCount;
-                        }));
+    private void addToResults(CompletionResultSet resultSet, List<PackageInRepo> packageInRepo, DependencyInfo key) {
+
+        Set<VersionResults> versionResults =
+                packageInRepo.stream().map(versionExplorer::getVersions).collect(Collectors.toSet());
+
+        List<CompletableFuture<?>> pendingFutures = versionResults.stream()
+                .filter(results -> results.versions().isEmpty())
```

Comment:
This is actually super strange to me (I guess I'm still thinking in the mindset where you give a `List<PackageInRepo>` to `VersionExplorer#getVersions`).

For each `getVersions` call, is it the case that it's _either_ some already loaded set of versions or it's a future to get more? If so, we should actually structure `VersionsResults` as an choice between the two somehow. I think you can do this really nicely in Java 21, but in Java 17 I think you can do something like:

```java
sealed interface VersionsResult permits AlreadyLoadedVersions, StillLoadingVersions {}

record AlreadyLoadedVersions(Set<DependencyVersion>)  {}
record StillLoadingVersions(CompletableFuture<Set<DependencyVersion>>)  {}
```

Reply:
I think there's definitely a type somewhere here that takes a set/list of `VersionsResult` and gives you methods like how many are still loading, how many versions are loaded etc

Reply:
Yeah it is the case that you either get the results or the future for the results - could we do something like this:

```java
    public CompletableFuture<Set<DependencyVersion>> getVersionsFutures(PackageInRepo packageInRepo) {
        String urlString = urlFor(packageInRepo);
        return shortLivedVersionCache.get(urlString);
    }
``` 

Then we could handle the futures on the completion side

Reply:
So I have made quite a few changes, I've switched `getVersions` to take a `List<PackageInRepo>` and now it returns a new type `VersionsResults` which handles the results aggregation. This makes `addToResults` alot neater as a lot of the logic is handled in the results type